### PR TITLE
Use react-components for open traces and views and related updates

### LINKED ISSUE #4: The trace tree in the trace explorer should show all openable traces
Now, only the top level directories of the workspace are shown and can be opened.

First step would be a smarter discovery of traces: go deep in the tree to locate potential traces and CTF folders and show the structure as a tree so that upper directories can be opened along with the traces underneath.

### LINKED ISSUE #7: Migrate support of trace/experiment opening
In the theia-trace-extension it is possible to select a directory or trace from the file explorer and execute "Open with -> Trace Viewer". After executing this the Trace Explorer is populated.

When no Trace is open, the "Open Trace" button is shown in the Trace Explorer and upon pressing the button the user can select a directory to find and open traces.

When executing that command, the command execution finds all CTF traces recursively in the directory and opens them in an experiment on the trace server. If the selection is a single file, then that file is opened as trace.

### LINKED ISSUE #11: Migrate the support of multiple open traces
It's possible to open multiple traces, but the available outputs don't update when switching the trace container, or selecting a trace in the opened traces.

### LINKED ISSUE #16: Migrate implementation to decouple delete experiments from trace-viewer
The change of PR https://github.com/theia-ide/theia-trace-extension/pull/248 needs to be applied in vscode as well.

### PR DESCRIPTION
**Pull request by bhufmann**
- This change includes updates to propagate signals between webviews.
This will allow views to be updated when one view changes. For example,
when the user selects an open trace in the "Traces" view, the selection
is sent to the various react-components. The "Available Views" view
will use that signal as trigger to fetch the available views from
the remote server.

- Remove analysis and traces tree view from the extension. They are
replaced by the theia-trace-extension react components and
corresponding webviews.

- The classes AnalysisTree and TracesTree is still in the code base to
allow opening traces from the file system using the context sensitive
menu and "Open in Trace Viewer" command. This will have to be clean-up
in a later patch (+ using a cancelable progress monitor).

- The patch also includes a fix to make source-map work when debugging
the webview apps.

- An additional commit to introduce ESLINT to the project and related code 
updates

Needs https://github.com/theia-ide/theia-trace-extension/pull/295

Fixes #7
Fixes #11
Fixes #16

Signed-off-by: Bernd Hufmann <Bernd.Hufmann@ericsson.com>

### CODE REVIEW

**Comment by bhufmann** [comment]
The following sequence diagram illustrates the communication between the webviews and the vscode extension using the example of selecting a trace. The communication between webviews and vscode extension is described in [1] and [2]. 

image


[1] https://code.visualstudio.com/api/extension-guides/webview#passing-messages-from-an-extension-to-a-webview 
[2] https://code.visualstudio.com/api/extension-guides/webview#passing-messages-from-a-webview-to-an-extension

File: vscode-trace-extension/src/trace-explorer/opened-traces/trace-explorer-opened-traces-view-provider.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
+import * as vscode from 'vscode';
```
Comment (tahini) [inline vscode-trace-extension/src/trace-explorer/opened-traces/trace-explorer-opened-traces-view-provider.ts:1]: I had to scroll back and forth a few time to catch the difference with the below `trace-explorer-views-provider.ts` file. Should that other file be `trace-explorer-views-view-provider.ts`? or replace the `view` in the name by `webview` and do the same below?

File: vscode-trace-webviews/src/style/trace-explorer.css [hunk 2]
```diff
@@ -0,0 +1,12 @@
+:root {
```
Comment (tahini) [inline vscode-trace-webviews/src/style/trace-explorer.css:1]: Should it be trace-viewer.css?

File: vscode-trace-webviews/src/trace-explorer/opened-traces/vscode-trace-explorer-opened-traces-widget.tsx [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+import '../../style/react-contextify.css';
+import { ExperimentManager } from '@trace-viewer/base/lib/experiment-manager';
+
+interface OpenedTracesAppState {
+  tspClientProvider: ITspClientProvider | undefined;
+}
+
+const MENU_ID = 'traceExplorer.openedTraces.menuId';
+
+class TraceExplorerOpenedTraces extends React.Component<{}, OpenedTracesAppState>  {
```
Comment (tahini) [inline vscode-trace-webviews/src/trace-explorer/opened-traces/vscode-trace-explorer-opened-traces-widget.tsx:20]: Do we still call the classes TraceExplorer even if the view is now called Trace Viewer?

**Review by tahini — CHANGES_REQUESTED** [review]
This is a _great_ improvement on the current situation! Thanks for this patch. My comments are very minor. See in the code. Also, I unfortunately have not followed closely recent progresses in theia-trace-extension, so sorry if I should know something that I don't.

1- The `Running the extension` part of the Readme does not apply anymore, especially the part about how to open a trace. Could you please update this section.

2- Also, now closing the trace does not delete it from the server anymore, right? We have to right-click and choose `Delete trace`?

**Comment by tahini** [comment]
Also, I don't think it fixes #4. This one will require that we parse the user's workspace to detect all possible traces. It is not in the scope of this PR.

File: vscode-trace-extension/src/trace-explorer/opened-traces/trace-explorer-opened-traces-view-provider.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+					} 
+					return;
+				case 'closeTrace':
+					if (message.data && message.data.experiment) {
+						TraceViewerPanel.disposePanel(this._extensionUri, message.data.experiment.name);
+					} 
+					return;
+				case 'deleteTrace':
+					if (message.data && message.data.experiment) {
+						// just remove th-e panel here
```
Comment (PatrickTasse) [inline vscode-trace-extension/src/trace-explorer/opened-traces/trace-explorer-opened-traces-view-provider.ts:71]: the

File: vscode-trace-extension/src/trace-explorer/opened-traces/trace-explorer-opened-traces-view-provider.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
+import * as vscode from 'vscode';
+import { TraceViewerPanel } from '../../trace-viewer-panel/trace-viewer-panel';
+import { getTspClientUrl, getTraceServerUrl } from '../../utils/tspClient';
+import { signalManager, Signals } from '@trace-viewer/base/lib/signals/signal-manager';
+import { Experiment } from 'tsp-typescript-client/lib/models/experiment';
+import { OpenedTracesUpdatedSignalPayload } from '@trace-viewer/base/lib/signals/opened-traces-updated-signal-payload';
+
+export class TraceExplorerOpenedTracesViewProvider implements vscode.WebviewViewProvider {
+
+	public static readonly viewType = 'traceExplorer.openedTracesView';
```
Comment (PatrickTasse) [inline vscode-trace-extension/src/trace-explorer/opened-traces/trace-explorer-opened-traces-view-provider.ts:10]: tab spacing, ok?

File: vscode-trace-webviews/src/trace-explorer/opened-traces/vscode-trace-explorer-opened-traces-widget.tsx [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        <Item id="open-id" onClick={this.handleItemClick}>Open Trace</Item>
+        <Item id="close-id" onClick={this.handleItemClick}>Close Trace</Item>
+        <Item id="delete-id" onClick={this.handleItemClick}>Delete Trace</Item>
+      </Menu>
+    </>
+    );
+  }
+
+  protected handleItemClick = (args: ItemParams) => {
+    console.log(args.event.currentTarget.id + ": " +(args.props.experiment as Experiment).UUID);
```
Comment (PatrickTasse) [inline vscode-trace-webviews/src/trace-explorer/opened-traces/vscode-trace-explorer-opened-traces-widget.tsx:117]: remove?

File: vscode-trace-webviews/src/trace-explorer/views/vscode-trace-explorer-views-widget.tsx [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+    this._signalHandler = new VsCodeMessageManager();
+    window.addEventListener('message', event => {
+
+      const message = event.data; // The JSON data our extension sent
+      switch (message.command) {
+        case 'set-tspClient':
+          this.setState({ tspClientProvider: new TspClientProvider(message.data) });
+          break;
+        case 'experimentSelected':
+          if(message.data) {
```
Comment (PatrickTasse) [inline vscode-trace-webviews/src/trace-explorer/views/vscode-trace-explorer-views-widget.tsx:40]: spacing

**Review by PatrickTasse — CHANGES_REQUESTED** [review]

File: vscode-trace-extension/src/trace-explorer/opened-traces/trace-explorer-opened-traces-view-provider.ts (hunk 1 above)
Comment (bhufmann) [reply vscode-trace-extension/src/trace-explorer/opened-traces/trace-explorer-opened-traces-view-provider.ts:1]: yeah, I struggle myself keeping track with the different files. I think adding "webview" makes totally sense and helps with clarity.

File: vscode-trace-webviews/src/style/trace-explorer.css (hunk 2 above)
Comment (bhufmann) [reply vscode-trace-webviews/src/style/trace-explorer.css:1]: I used the corresponding name as in theia-trace-extension where also some clean-up needs to be done. Makes sense to change it here. Please note that the styling in this patch works for now. But some updates are required in the theia-trace-extension's react-components to support better styling and theming independent of the "theia" reference in the variables.

File: vscode-trace-webviews/src/trace-explorer/opened-traces/vscode-trace-explorer-opened-traces-widget.tsx (hunk 3 above)
Comment (bhufmann) [reply vscode-trace-webviews/src/trace-explorer/opened-traces/vscode-trace-explorer-opened-traces-widget.tsx:20]: Good point. I'm aware of it. In the theia-trace-extension we have the same mismatch. The user facing strings were changed from "Trace Explorer" to "Trace Viewer", but the classes and variables stayed the same. To not introduce more confusion, I kept the same discrimination in the vscode-trace-extension than in the theia-trace-extension. But yeah, a code clean-up should be done in both repos.

IMHO, the discrimination between "Trace Explorer" for traces/analysis and "Trace Viewer" for the widgets part made sense, but the majority where against it.

**Comment by bhufmann** [comment]
> Also, I don't think it fixes #4. This one will require that we parse the user's workspace to detect all possible traces. It is not in the scope of this PR.

After re-reading the #4, I realized that it describes something else, i.e. trace detection on the file system. My understanding was that it was showing the available traces on the server. I'll remove it.

File: vscode-trace-extension/src/trace-explorer/opened-traces/trace-explorer-opened-traces-view-provider.ts (hunk 4 above)
Comment (bhufmann) [reply vscode-trace-extension/src/trace-explorer/opened-traces/trace-explorer-opened-traces-view-provider.ts:71]: Done

File: vscode-trace-extension/src/trace-explorer/opened-traces/trace-explorer-opened-traces-view-provider.ts (hunk 5 above)
Comment (bhufmann) [reply vscode-trace-extension/src/trace-explorer/opened-traces/trace-explorer-opened-traces-view-provider.ts:10]: Yeah, we should add a linter to catch that and update all classes to comment code style settings. Outside of the scope this patch. I'll update here though.

File: vscode-trace-webviews/src/trace-explorer/views/vscode-trace-explorer-views-widget.tsx (hunk 7 above)
Comment (bhufmann) [reply vscode-trace-webviews/src/trace-explorer/views/vscode-trace-explorer-views-widget.tsx:40]: Done

File: vscode-trace-extension/src/trace-explorer/opened-traces/trace-explorer-opened-traces-view-provider.ts (hunk 5 above)
Comment (bhufmann) [reply vscode-trace-extension/src/trace-explorer/opened-traces/trace-explorer-opened-traces-view-provider.ts:10]: I added a 2nd commit for eslint usage and related updates

**Comment by bhufmann** [comment]
https://github.com/theia-ide/theia-trace-extension/pull/295 has been merged to master.

**Review by PatrickTasse — APPROVED** [review]

**Review by PatrickTasse — APPROVED** [review]

**Comment by bhufmann** [comment]
I also updated the readme and added new screenshots.

**Review by PatrickTasse — APPROVED** [review]

**Comment by bhufmann** [comment]
@tahini  are ok with this PR as well?

**Review by tahini — APPROVED** [review]
Thanks!

**Comment by bhufmann** [comment]
Thanks for the reviews!

**Pull request merged**